log/file.go

```diff
@@ -0,0 +1,157 @@
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package log
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+// defaultLogfilePath is a default logfile path.
+const defaultLogfilePath = "/tmp/k6run.log"
+
+// fileHook is a hook to handle writing to local files.
+type fileHook struct {
+	fallbackLogger logrus.FieldLogger
+	ch             chan []byte
+	path           string
+	w              io.WriteCloser
+	levels         []logrus.Level
+}
+
+// FileHookFromConfigLine returns new fileHook hook.
+func FileHookFromConfigLine(
+	ctx context.Context, fallbackLogger logrus.FieldLogger, line string,
+) (logrus.Hook, error) {
+	hook := &fileHook{
+		fallbackLogger: fallbackLogger,
+		ch:             make(chan []byte),
+		path:           defaultLogfilePath,
+		levels:         logrus.AllLevels,
+	}
+
+	if fallbackLogger != nil {
+		hook.fallbackLogger = fallbackLogger
+	}
+
+	if line != "file" {
+		parts := strings.SplitN(line, "=", 2)
+		if parts[0] != "file" {
+			return nil, fmt.Errorf("logfile configuration should be in the form `file=path-to-local-file` but is `%s`", line)
+		}
+
+		if err := hook.parseArgs(line); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := hook.createOpenFile(); err != nil {
+		return nil, err
+	}
+
+	go hook.loop(ctx)
+
+	return hook, nil
+}
+
+func (h *fileHook) parseArgs(line string) error {
+	tokens, err := tokenize(line)
+	if err != nil {
+		return fmt.Errorf("error while parsing logfile configuration %w", err)
+	}
+
+	for _, token := range tokens {
+		switch token.key {
+		case "file":
+			if token.value == "" {
+				return fmt.Errorf("filepath must not be empty")
+			}
+			h.path = token.value
+		case "level":
+			h.levels, err = parseLevels(token.value)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown logfile config key %s", token.key)
+		}
+	}
+
+	return nil
+}
+
+// createOpenFile opens logfile by provided path,
+// it will create a file if it doesn't exist.
+func (h *fileHook) createOpenFile() error {
+	dir := filepath.Dir(h.path)
+	if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+		return fmt.Errorf("failed to create dir for logfile %s: %w", h.path, err)
+	}
+
+	file, err := os.OpenFile(h.path, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0o600)
+	if err != nil {
+		return fmt.Errorf("failed to open logfile %s: %w", h.path, err)
+	}
+
+	h.w = file
+	return nil
+}
+
+func (h *fileHook) loop(ctx context.Context) {
```

Comment:
A nice pattern I discovered when reading the Concurrency in Go book, is to "confine" concurrency: confining concurrency to a given scope in order to avoid data races. Following this pattern, we would refactor this code to look like this:

```go
func loop(ctx context.Context) chan []byte {
	ch := make(chan []byte)

	go func() {
		defer close(ch)

		for {
			select {
			case entry := <-h.ch:
				if _, err := h.w.Write(entry); err != nil {
					h.fallbackLogger.Errorf("failed to write a log message to a logfile: %w", err)
				}
			case <-ctx.Done():
				if err := h.w.Close(); err != nil {
					h.fallbackLogger.Errorf("failed to close logfile: %w", err)
				}

				return
			}
		}
	}()
	
	return ch
}
```

And in the constructor we would then do something along the lines of:
```go
hook.ch = hook.loop(ctx)
```

I need to explicit that this is in no way a request. When seeing this code, and having to make a bit of mental gymnastic to figure out who declared/used/closed the `ch`, I thought it might be worth mentioning this pattern here. Let me know what you think :smile_cat:

Reply:
That's an interesting pattern. Thanks for the recommendation!
I applied it 🙂